src/utils/log_redact.py

```
@@ -0,0 +1,307 @@
+        # (custom ``Exception`` subclass, buggy ``__str__`` raising
+        # ``RuntimeError`` / ``AttributeError``, etc.) must NOT escape
+        # the filter — we're attached via ``setLogRecordFactory`` so this
+        # runs on EVERY ``logger.*()`` call, and escalating here would
+        # break the caller's normal execution including records that
+        # would otherwise be dropped later by level filtering. Every
+        # blind catch below leaves the record untouched and returns
+        # ``True`` so ``logging`` surfaces the error via its own handler.
+        if record.args:
+            try:
+                formatted = record.getMessage()
+            except Exception:
```

Ruff (0.15.10)

[warning] Do not catch blind exception: `Exception`

(BLE001)